docs/development/ADRs/0017-Toolchain-Configuration.md

```diff
@@ -0,0 +1,101 @@
+
+## Context
+
+In this document the word toolchain is used to mean all the code components that work together to go from DSL code to an optimized, runnable python callable.
+It includes JIT / OTF pipelines or workflows but also transformation passes, lowerings, parsers etc.
+
+The most pressing issue is the developer experience. One debugging technique is to have the generated C++ code written to a permanent file location for inspection. This requires changing the code to reconfigure the build cache location. However this is forseeably only one of multiple values that stencil developers and their end users will wish to configure without touching GT4Py code.
+
+At the time of creation of this document, at least one additional toolchain is actively being worked on and there are plans to make additional parts of the existing toolchains configurable, both of which will compound the issue.
+
+**Concerns**
+
+- Some ways of configuring toolchains involve configuring multiple components in synch
+- Some ways of configuring toolchains involve switching out or nesting toolchain steps
+- What configuration options are avaiable will depend on what toolchain will be used and how that is configured.
+- Hierarchical configuration defaults and overrides can be confusing from a user perspective.
+- Leaving the configuration interface completely up to toolchain developers could lead to a confusing ad fragmented user experience.
+
+## Decision
+
+### Opt-in pattern for building toolchains from user configuration and client code
+
+Any toolchain that has user configurable options should provide a high level interface for building a toolchain that is consistent with the options set by the end user. If a default toolchain instance is provided in GT4Py code, it should use that interface. This ensures that the simplest way of obtaining an instance of a toolchain respects user configuration.
+
+### Limit configuration options exposed to the end user
+
+Any option that the end user can change in order to influence the toolchain behavior must be defined in `gt4py.next.config` with
+
+- an internal name (a module level variable)
+- an external name used to load from environment variables (possibly with a common prefix)
+- a fallback default value in case no environment variable is defined
+
+Any other toolchain option is considered an implementation detail.
+
+### Limit the times when configuration can change
+
+By making the `gt4py.next.config` module contain module level variables with user configuration, we ensure user configuration can only be changed between python interpreter runs (after `from gt4py import next` the configuration is fixed). Of course there are ways around it but they should be considered unsupported as they are difficult to make reliable (consider monkey patching as an example).
+
+### Environment variables are the primary end user interface
```

Comment:
All of these are choices, nothing is needed. The only choice which is fairly well justified is to not try to warn users of "rogue" toolchains, because most of the cycle went into experimentation for that.

Reply:
I have hopefully made it more clear that this is not a requirement, but a decision that was made for now. I have outlined a sketch of what the alternative could look like (in the alternatives section).